Feature/pipe pending handles 2

indutny opened the pull request:
Github isn't updating previous PR because it was closed.

/cc @jclulow @bnoordhuis @saghul

indutny committed:
pipe: allow queueing pending handles

Introduce `uv_pipe_has_pending_handles()`, may be called in `read2_cb`
to accept all pending incoming handles on IPC pipe:

    do
      uv_accept(...);
    while (uv_pipe_has_pending_handles(pipe));

indutny committed:
api tweak

indutny committed:
better api

indutny committed:
make uv__handle_type static again

indutny committed:
fixes

indutny committed:
realloc improvement

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit indutny/libuv@57275e24d484ed05993f663900f4bc9afc71ddfe has the following error(s):
- Commit message must indicate the subsystem this commit changes

Commit indutny/libuv@f159f684f78c74ff4ac2a1c10665d72019b9c571 has the following error(s):
- Commit message must indicate the subsystem this commit changes

Commit indutny/libuv@339fda6d175794e793363697e671b32ac50f9333 has the following error(s):
- Commit message must indicate the subsystem this commit changes

Commit indutny/libuv@4a8dcb8b63abdcd9c50ce1996a540f8833f54978 has the following error(s):
- Commit message must indicate the subsystem this commit changes

Commit indutny/libuv@179877d2ab931cc7db50362b7af916ad6080ebcb has the following error(s):
- Commit message must indicate the subsystem this commit changes

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

indutny commented:
<blockquote>
Put another way, it scans for the SCM_RIGHTS message but when it finds it, it only considers the first data element. It needs to check cmsg->cmsg_len
</blockquote>

@bnoordhuis - good point, I'll make walk all fds in cmsg tomorrow.

indutny committed:
stream: accept all incoming fds

indutny commented:
@bnoordhuis addressed that thing.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -56,6 +56,14 @@ struct uv__stream_select_s {
+
+struct uv__stream_queued_fds_s {
+  int size;
+  int offset;
```
Don't use signed types for fields that always contain values >= 0.

txdv replied:
Size is always a positive too

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -586,9 +595,33 @@ int uv_accept(uv_stream_t* server, uv_stream_t* client) {
+    server->accepted_fd = queued_fds->fds[0];
+
+    /* All read, free */
+    if (--queued_fds->offset == 0) {
```
Should probably be preceded by an assert that checks that queued_fds->offset > 0.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -586,9 +595,33 @@ int uv_accept(uv_stream_t* server, uv_stream_t* client) {
+    }
+
+    /* Invoke read_cb one more time */
+    uv__io_feed(server->loop, &server->io_watcher);
```
Borderline incorrect: uv__io_feed() queues a UV__POLLOUT event but a server handle is only primed for UV__POLLIN.

indutny replied:
Oh gosh, you're right.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -951,6 +984,46 @@ static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
+    queued_fds = malloc((queue_size - 1) * sizeof(*queued_fds->fds) +
+                        sizeof(*queued_fds));
+    if (queued_fds == NULL)
+      return UV_ENOMEM;
```
Convention is to return -ENOMEM.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -951,6 +984,46 @@ static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
+     * Allocation failure, report back.
+     * NOTE: if it is fatal - sockets will be closed in uv__stream_close
+     */
+    if (queued_fds == NULL) {
```
Leaks memory.

indutny replied:
No, it doesn't

bnoordhuis replied:
Oh, you're right.  I can see it now, I just discovered GH's new expand functionality.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -958,6 +1031,17 @@ static void uv__read(uv_stream_t* stream) {
+    stream->read2_cb((uv_pipe_t*) stream,
+                     0,
+                     &buf,
+                     uv__handle_type(stream->accepted_fd));
```
Seems like a terrible change because you're putting the burden on the user to always check if buf.base == NULL.  It's unlikely that existing code does that, so you'll be breaking pretty much everything that's already out there.

It's made worse because receiving multiple file descriptors is an unlikely event so any bugs will be hidden for a long time and difficult to reproduce.  If you put together this PR for your own convenience, then I suggest you drop it.

indutny replied:
It was reproducible every time when I did C clustering benchmarks on a smartos server. The only alternative to that is to ACK all incoming fds in a child processes, which is a bit of burden and waste of resources.

bnoordhuis replied:
I don't understand what you mean.  My point is that you're changing the API in a subtle way that's bound to break existing programs.  That's not a good thing.

indutny replied:
It can already be a null-buffer, when nread is UV_EOF

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -1046,17 +1130,24 @@ static void uv__read(uv_stream_t* stream) {
+            void* pv = CMSG_DATA(cmsg);
+            int* pi = pv;
+            unsigned int i;
+            unsigned int count = (cmsg->cmsg_len - CMSG_LEN(0)) / sizeof(*pi);
```
You cannot depend on count actually being the number of file descriptors in the message.  As I mentioned in the other PR, it doesn't work that way on OS X (and it's probably the same on FreeBSD) - it reports the _original_ number of file descriptors, which may be more than what fits in the receive buffer.

indutny replied:
Would using `msg_controllen` fix it?

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -1046,17 +1130,24 @@ static void uv__read(uv_stream_t* stream) {
+                err = uv__stream_queue_fd(stream, pi[i]);
+                if (err != 0) {
+                  uv__stream_read_cb(stream, err, NULL, UV_UNKNOWN_HANDLE);
+                  return;
```
Leaks the remaining file descriptors.

You should probably set the close-on-exec flag on the file descriptors first.  Linux has a MSG_CMSG_CLOEXEC flag that you can pass to recvmsg() which does that atomically but it wasn't added until 2.6.22 or 2.6.23.  You'll need a fallback path for when recvmsg() returns EINVAL.

indutny replied:
I think I can just close them.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -1493,6 +1587,15 @@ void uv__stream_close(uv_stream_t* handle) {
+  if (handle->queued_fds != NULL) {
+    queued_fds = handle->queued_fds;
+    for (i = 0; i < queued_fds->offset; i++)
+      close(queued_fds->fds[i]);
```
uv__close()

bnoordhuis commented:
If you go through with this, it _really_ needs some regression tests.  Should be easy: create file descriptors with socketpair(), feed one to uv_pipe_open() and call sendmsg() on the other.  You'll want to cover several cases: one file descriptor, several and more than fit in the receive buffer.  Keep in mind that OpenBSD will fail with EMSGSIZE if number_of_fds > sizeof_receive_buffer.

indutny committed:
fixes

indutny committed:
unix: `uv__io_feed(loop, watcher, events)`

Add third argument to the `uv__io_feed`, specifying which event it
should "feed" (either `UV__POLLIN`, or `UV__POLLOUT`).

indutny commented:
@bnoordhuis all fixed, working on a test.

indutny committed:
test

indutny commented:
@bnoordhuis added test. @saghul please review it too!

saghul commented on src/unix/stream.c:
```diff
@@ -951,13 +985,112 @@ static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
+  int err;
+
+  /*
+   * Some implementations can send multiple file descriptors in a
```
I see you use CMSG_NXTHDR, is this comment still relevant?

indutny replied:
No, I'll remove it.

saghul commented:
I had a look at it, but TBH I don't think I know enough about cmsg\* to make any meaningful comment.

re: passing a NULL buf: in pyuv I assume the buffer base is not NULL if nread is >= 0, so this code would break my assumption. Not sure about what others do. Maybe the buffer should contain the empty string instead?

indutny commented:
@saghul the zero-length buffer won't work and may break existing things too, considering that users may think that the buffer has came from `alloc_cb`.

saghul commented:
Ah, indeed. Hum I don't see how we can get away with this without "breaking" something :-S

indutny commented:
Yeah... Let's try discussing it this weekend or next week on irc. I'll write a message on the mailing list to get some feedback from users.

indutny committed:
wip

indutny committed:
Revert "unix: `uv__io_feed(loop, watcher, events)`"

This reverts commit 674b0736bcb7b758f3f53a02a67a0a0dfa7325dd.

indutny committed:
do not feed

indutny committed:
unix support

indutny committed:
win

indutny committed:
pipe: fixes

indutny committed:
fix

indutny commented:
@jclulow @bnoordhuis @saghul done

jclulow commented on include/uv-win.h:
```diff
@@ -424,10 +424,11 @@ typedef int (WSAAPI* LPFN_WSARECVFROM)
   uv_write_t ipc_header_write_req;                                            \
   int ipc_pid;                                                                \
   uint64_t remaining_ipc_rawdata_bytes;                                       \
-  unsigned char reserved[sizeof(void*)];                                      \
```
What was this reserved space for?  Is removing it OK?

indutny replied:
It is safe to remove it in unstable version. According to b496c1221c3816f06a3fcd3e8c3efedcf6eb8445 it was introduced for a future feature implementation in a stable branch.

@piscisaureus wanna weigh in?

piscisaureus replied:
That used to be another field that we removed on a stable branch. So we added a reserved spot to not break the ABI
(I see now that replacing `void*` by `char[sizeof void]` only works accidentally but was not right.)

piscisaureus commented on src/win/pipe.c:
```diff
@@ -31,6 +31,17 @@
+
+struct uv__ipc_queue_item_s {
+  /*
+   * NOTE: It is important for socket_info to be the first field,
```
I don't get this. You mean you're going to assign a `struct uv__ipc_queue_item_s*` to a `WSAPROTOCOL_INFOW*` ?
Why not use container_of (or what have you on windows) then?

indutny replied:
No, because it'll be passed to `free()`.

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,11 +373,22 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
         handle->pending_ipc_info.socket_info = NULL;
       }
 
+      /* Free rest pending sockets */
```
Grammar...

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,11 +373,22 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
       }
 
+      /* Free rest pending sockets */
+      /* XXX: Need to free sockets themselves? */
```
Yes you need to "materialize" the socket with `WSASocket`, then close it with `CloseSocket`.

piscisaureus commented on src/win/pipe.c:
```diff
@@ -1345,6 +1372,34 @@ static void uv_pipe_read_error_or_eof(uv_loop_t* loop, uv_pipe_t* handle,
+  WSAPROTOCOL_INFOW* socket_info;
+  uv__ipc_queue_item_t* item;
+
+  if (handle->pending_ipc_info.socket_info == NULL) {
```
I don't agree with having both a queue _and_ having the queue 'head' data stored in separate fields.
Please drop the `handle->pending_ipc_info` field and replace it with a queue head.

indutny replied:
It is somewhat complicated if we don't want to export the `uv__ipc_queue_item_t` declaration. Also this was done to make it work without allocations in the most of the cases.

piscisaureus commented on src/win/pipe.c:
```diff
@@ -1745,3 +1778,21 @@ int uv_pipe_open(uv_pipe_t* pipe, uv_file file) {
+  if (handle->pending_ipc_info.socket_info == NULL)
+    return 0;
+
+  return 1 + handle->pending_ipc_info.queue_len;
```
And with `handle->pending_ipc_info` being replaced by a queue, you don't have to add 1 here.

indutny replied:
Please let me know you have any suggestions for embedding internal struct in a public one.

indutny committed:
fix grammar

jclulow commented on src/unix/stream.c:
```diff
@@ -552,6 +552,7 @@ void uv__server_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
 
 int uv_accept(uv_stream_t* server, uv_stream_t* client) {
   int err;
+  uv__stream_queued_fds_t* queued_fds;
```
This variable doesn't seem to be used outside of the `if` block starting at L591.  It would be clearer if it were declared down there.

trevnorris replied:
Doesn't libuv follow the C89 spec? Part of which requires all variables to
be declared at the top of the function.

jclulow commented on src/unix/stream.c:
```diff
@@ -931,23 +953,102 @@ static uv_handle_type uv__handle_type(int fd) {
+
+static int uv__stream_queue_fd(uv_stream_t* stream, int fd) {
+  uv__stream_queued_fds_t* queued_fds;
+  unsigned int queue_size;
```
It's confusing to use `queue_size` for the `malloc` case, but not the `realloc` case.  Either move it inside the `if`, or use it consistently for both branches.

indutny replied:
Decided to use it in both branches

jclulow commented on src/unix/stream.c:
```diff
@@ -931,23 +953,102 @@ static uv_handle_type uv__handle_type(int fd) {
+     * NOTE: if it is fatal - sockets will be closed in uv__stream_close
+     */
+    if (queued_fds == NULL) {
+      queued_fds->size -= 8;
```
If `queue_fds == NULL` here, surely this will segfault.

indutny committed:
fix bug

indutny committed:
fixes

indutny commented:
@jclulow I think I have fixed all nits, thank you! Could you please take a look?

indutny committed:
minimize

indutny commented:
@piscisaureus pushed windows update

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,10 +370,30 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
+        item = QUEUE_DATA(q, uv__ipc_queue_item_t, member);
+
+        /* Materialize socket and close it */
+        socket = WSASocketW(AF_INET,
```
That should be `FROM_PROTOCOL_INFO`

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,10 +370,30 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
+
+        /* Materialize socket and close it */
+        socket = WSASocketW(AF_INET,
+                            SOCK_STREAM,
```
ditto

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,10 +370,30 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
+        /* Materialize socket and close it */
+        socket = WSASocketW(AF_INET,
+                            SOCK_STREAM,
+                            IPPROTO_IP,
```
ditto

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,10 +370,30 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
-        free(handle->pending_ipc_info.socket_info);
-        handle->pending_ipc_info.socket_info = NULL;
+      /* Free rest of the pending sockets */
+      /* XXX: Need to free sockets themselves? */
```
Obsolete comment

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,10 +370,30 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
-      if (handle->pending_ipc_info.socket_info) {
-        free(handle->pending_ipc_info.socket_info);
-        handle->pending_ipc_info.socket_info = NULL;
+      /* Free rest of the pending sockets */
```
s/rest of//

piscisaureus commented on src/win/pipe.c:
```diff
@@ -358,10 +370,30 @@ void uv_pipe_endgame(uv_loop_t* loop, uv_pipe_t* handle) {
+                            WSA_FLAG_OVERLAPPED);
+        free(item);
+
+        /* XXX: Handle error? */
```
I don't think handling the error here is necessary.

indutny committed:
fixes

indutny commented:
@piscisaureus thank you, all fixed.

indutny commented:
@jclulow LGTY?

piscisaureus commented:
LGTM

trevnorris commented on src/unix/stream.c:
```diff
@@ -586,9 +585,33 @@ int uv_accept(uv_stream_t* server, uv_stream_t* client) {
+done:
+  /* Process queued fds */
+  if (server->queued_fds != NULL) {
+    uv__stream_queued_fds_t* queued_fds;
```
@piscisaureus Doesn't libuv follow the C89 spec and require all variables to be declared at the top of the function?

bnoordhuis replied:
That's correct C89.  Variable declarations need to go at the top of a block which can be a function but doesn't have to be.

trevnorris replied:
Ah OK. Thanks. For some reason though it must be the function block. Guess
that's what I get for being a JS dev. :-P

jclulow commented on src/unix/stream.c:
```diff
@@ -931,23 +954,101 @@ static uv_handle_type uv__handle_type(int fd) {
+   */
+
+  for (cmsg = CMSG_FIRSTHDR(msg);
+       msg->msg_controllen > 0 && cmsg != NULL;
```
The check for `msg->msg_controllen > 0` is not necessary, as I understand -- it is included in `CMSG_FIRSTHDR` .  In fact, it's a more complete check in there, anyway: that `msg_controllen` is not just non-zero, but long enough to contain a valid `struct cmsghdr`.

jclulow commented on src/unix/stream.c:
```diff
@@ -931,23 +954,101 @@ static uv_handle_type uv__handle_type(int fd) {
+      continue;
+    }
+
+    /* silence aliasing warning */
```
What warning are you actually getting, here?  Can you not just `pi = (int *)CMSG_DATA(cmsg);` ?

indutny replied:
Actually, I copied it from another place :) Decided to keep it as it is.

I think it is done this way to remove `-fstrict-aliasing` warnings.

jclulow commented on src/unix/stream.c:
```diff
@@ -977,21 +1081,20 @@ static void uv__read(uv_stream_t* stream) {
     } else {
-      assert(stream->read2_cb);
-      /* read2_cb uses recvmsg */
+      /* ipc uses recvmsg */
```
So, looking at various things around the place, it feels like `msg_controllen` should be set with the `CMSG_LEN()` macro, having previously allocated memory of size returned by the `CMSG_SPACE()` macro.  i.e. depending on how many file descriptors we wish to be able to receive in a single message:

``` c
#define   MAX_RECV_FD_COUNT     16
#define   SCM_RIGHTS_LENGTH     (MAX_RECV_FD_COUNT * sizeof (int))

char cmsg_space[CMSG_SPACE(SCM_RIGHTS_LENGTH)];

cmsg.cmsg_len = msg.msg_controllen = CMSG_LEN(SCM_RIGHTS_LENGTH);
```

saghul replied:
Did you check this @indutny ?

indutny replied:
Yes, I think I fixed it. Tried it on Linux and Mac, was working fine.

jclulow commented on src/unix/stream.c:
```diff
@@ -931,23 +954,101 @@ static uv_handle_type uv__handle_type(int fd) {
+    /* silence aliasing warning */
+    pv = CMSG_DATA(cmsg);
+    pi = pv;
+    count = (msg->msg_controllen - CMSG_LEN(0)) / sizeof(*pi);
```
I believe you actually want to use `CMSG_DATA(cmsg)` here, rather than `CMSG_LEN(0)`.  These macros seem to do alignment stuff on different platforms, so it seems unwise to assume that the data region begins immediately following the end of the `struct cmsg_hdr` without padding.

indutny committed:
fixes

indutny commented:
@jclulow fixed nits, please take another look.

indutny committed:
stream: hack for linux

indutny commented:
@jclulow also pushed a hack for linux, it seems that it is adding a lot of alignment to `CMSG_LEN()`.

indutny commented:
Please review ;)

indutny commented:
Seems to be failing on linux right now, will figure it out tomorrow.

indutny committed:
better linux fix

indutny committed:
improvements

indutny committed:
fix pointer arithmetic

saghul commented on src/unix/pipe.c:
```diff
@@ -214,3 +214,27 @@ void uv_pipe_connect(uv_connect_t* req,
+int uv_pipe_pending_count(uv_pipe_t* handle) {
+  uv__stream_queued_fds_t* queued_fds;
+
+  assert(handle->ipc);
```
Can we be a bit more gentle here and return 0 instead?

saghul commented on src/unix/pipe.c:
```diff
@@ -214,3 +214,27 @@ void uv_pipe_connect(uv_connect_t* req,
+
+
+uv_handle_type uv_pipe_pending_type(uv_pipe_t* handle) {
+  assert(handle->ipc);
```
Ditto, but with UV_UNKNOWN_HANDLE.

indutny committed:
fix alignment problems

indutny committed:
fixes

indutny commented:
@saghul done!

indutny commented:
Also fixed alignment issues, seems to be working fine on linux.

indutny commented:
@jclulow ^

saghul commented:
@indutny sweet! I can't comment much on the rest as it's beyond my knowledge, so I'll leave it up to you guys :-)

indutny commented:
@jclulow ping.

saghul commented:
@indutny, please close https://github.com/joyent/libuv/issues/461 when you land this.

indutny commented:
@jclulow reminding about it again :)

saghul commented on include/uv.h:
```diff
@@ -1148,6 +1129,15 @@ UV_EXTERN void uv_pipe_connect(uv_connect_t* req, uv_pipe_t* handle,
 UV_EXTERN void uv_pipe_pending_instances(uv_pipe_t* handle, int count);
 
+/*
+ * Should be used to receive handles on ipc pipe.
```
"Used to receive handles over ipc pipes."

saghul commented on src/unix/pipe.c:
```diff
@@ -214,3 +214,31 @@ void uv_pipe_connect(uv_connect_t* req,
+    return 1;
+
+  queued_fds = handle->queued_fds;
+  return 1 + queued_fds->offset;
```
return handle->queued_fds->offset + 1 ?

saghul commented on src/unix/stream.c:
```diff
@@ -931,23 +954,105 @@ static uv_handle_type uv__handle_type(int fd) {
+    queue_size = queued_fds->size + 8;
+    queued_fds = realloc(queued_fds,
+                         (queue_size - 1) * sizeof(*queued_fds->fds) +
+                             sizeof(*queued_fds));
```
Align with realloc open parens? (maybe GH shows it weird)

indutny replied:
Actually, this is intended 4 space indent here, as it is not a new argument, but a continuation of expression on the previous line.

indutny commented:
Landed in b05a3ee, thank you!
